File: binary_search.rb
def bsearch(arr, target)
  return "NOT FOUND" if arr.empty?

  midpoint = arr.length / 2
  mid_val = arr[midpoint]

  if mid_val == target
    return midpoint
  elsif mid_val > target
    bsearch(arr[0...midpoint], target)
  elsif mid_val < target
    answer = bsearch(arr[(midpoint + 1)..-1], target)
    return "NOT FOUND" unless answer != "NOT FOUND"
    midpoint + 1 + answer
  end
end


arr = [1, 2, 4, 6, 9, 11, 16, 19, 24]
bsearch(arr, 24)

def bsearch_with_duplicates(arr, target)
  return if arr.empty?

  midpoint = arr.length / 2
  mid_val = arr[midpoint]

  if mid_val == target
    # FIND LEFTMOST ELEMENT
    # possible_left = bsearch_with_duplicates(arr[0...midpoint], target)
    # possible_left ? possible_left : midpoint

    # FIND RIGHTMOST ELEMENT
    possible_right = bsearch_with_duplicates(arr[(midpoint + 1)..-1], target)
    possible_right ? (possible_right + midpoint + 1) : midpoint

  elsif mid_val > target
    bsearch_with_duplicates(arr[0...midpoint], target)
  elsif mid_val < target
    answer = bsearch_with_duplicates(arr[(midpoint + 1)..-1], target)
    return unless answer
    midpoint + answer + 1
  end
end

arr_with_dups = [1, 1, 3, 3, 3, 6, 6, 6, 6, 6, 9, 9, 9, 11, 15, 17, 18, 18]
p bsearch_with_duplicates(arr_with_dups, 3)

File: binary_search.js
const binarySearch = (arr, targ) => {
  if (arr.length === 0) return "NOT FOUND";

  const midpoint = Math.floor(arr.length / 2);
  const midVal = arr[midpoint];

  if (midVal === targ) {
    return midpoint;
  } else if (midVal > targ) {
    return binarySearch(arr.slice(0, midpoint), targ);
  } else {
    const answer = binarySearch(arr.slice(midpoint + 1), targ);
    if (answer === "NOT FOUND") return "NOT FOUND";

    return midpoint + answer + 1;
  }

};

const arr = [1, 2, 6, 7, 9, 12, 33, 56, 76, 888, 1234];

console.log(binarySearch(arr, 76));
